Update TF fine-tuning docs (#18654)
huggingface/transformers: docs/source/en/training.mdx

<a id='trainer'></a>

<frameworkcontent>
<pt>
<Youtube id="nvBXf7s7vTI"/>

> ## Train

Reviewer: Train needs to be outside of the framework-content block.

At this point, you should follow the section corresponding to the framework you want to use. You can use the links
in the right sidebar to jump to the one you want - and if you want to hide all of the content for a given framework,
just use the button at the top-right of that framework's block!

> 

Reviewer: Same for this intro.

Reviewer 2: Love this little intro for guiding user navigation/improving readability!

<a id='trainer'></a>

> ## Train with PyTorch Trainer

Reviewer: The Trainer is not from PyTorch, so this should be more like
```suggestion
## Train your PyTorch model with the Trainer
```

> ## Train your PyTorch model with the Trainer

Reviewer: No this section groups both training with PyTorch and TensorFlow with framework-specific blocks, so leave this as it was.

<Youtube id="rnTGBy2ax1c"/>

## Train with Keras

🤗 Transformers models also supports training in TensorFlow with the Keras API.

### Loading data for Keras

When you want to train a 🤗 Transformers model with the Keras API, you need to convert your dataset to a format that
Keras understands. If your dataset is small, you can just convert the whole thing to NumPy arrays and pass it to Keras.

First, load a dataset:

```py
from datasets import load_dataset

dataset = load_dataset("glue", "cola")  # Simple text classification dataset

> dataset = dataset["train"]  # Just take the training split for now

Reviewer: Use the prose to explain what you're doing instead of comments in the code.

Author: Done!

```

Next, load a tokenizer and tokenize the data as NumPy arrays:

```py
from transformers import AutoTokenizer

tokenizer = AutoTokenizer.from_pretrained("bert-base-cased")
tokenized_data = tokenizer(dataset["text"], return_tensors="np", padding=True)

> labels = np.array(dataset["label"])  # Label is already an array of 0 and 1

Reviewer: Same here.

Author: Done!

<Youtube id="rnTGBy2ax1c"/>

> ## Train with Keras

Reviewer: Same for this line.

# Load and compile our model
model = TFAutoModelForSequenceClassification.from_pretrained("bert-base-cased")
# Lower learning rates are often better for fine-tuning transformers
model.compile(optimizer=Adam(3e-5))

model.fit(tokenized_data, labels)
```

<Tip>

Note that we don't pass a loss argument to our models when we `compile()` them! Hugging Face models will automatically
choose a loss that is appropriate for their task and model architecture if this argument is left blank. You can always
override this by specifying a loss yourself if you want to!

</Tip>

This approach is great when it works, but for larger datasets you might find it starting to become a problem. Why?
Because the tokenized array and labels would have to be fully loaded into memory, and because Numpy doesn’t handle
“jagged” arrays, so every tokenized sample would have to be padded to the length of the longest sample in the whole
dataset. That’s going to make your array even bigger, and all those padding tokens will slow down training too!

### Loading data as a tf.data.Dataset

If you want to avoid this, you can load your data as a `tf.data.Dataset` instead. Although you can write your own

> `tf.data` pipeline if you want, we have two convenience methods for doing this - [`~datasets.Dataset.to_tf_dataset`]

Author: Done!

and [`~TFPreTrainedModel.prepare_tf_dataset`]. We recommend `prepare_tf_dataset()` in most cases because it is simpler
and can automatically infer which columns from the dataset to load by comparing them against the input names for the
model. The `to_tf_dataset()` method is more low-level, and is useful when you want to exactly control how your
dataset is created.

Before you can use this method, you will need to add the tokenizer outputs to your dataset as columns, as in
the following code sample:

```py
def tokenize_dataset(data):
    # Keys of the returned dictionary will be added to the dataset as columns
    return tokenizer(data["text"])

dataset = dataset.map(tokenize_dataset)
```

Remember that Hugging Face datasets are stored on disk by default, so this will not inflate your memory usage! Once the
columns have been added, we can now stream batches from the dataset and add padding on the fly, which greatly
reduces the number of padding tokens compared to padding the entire dataset.

```py
>>> tf_dataset = model.prepare_tf_dataset(dataset, batch_size=16, shuffle=True, tokenizer=tokenizer)
```

<Tip>

> The reason we pass the tokenizer to `prepare_tf_dataset` is so that it can correctly pad batches as they're loaded.

Reviewer 2: This is a pretty substantial paragraph, so I think it's better to include it in the main body of content instead of as a Tip.

```suggestion
You need to pass the tokenizer to `prepare_tf_dataset` so it can correctly pad batches as they're loaded.
```

Author: Done! I also included those edits.

If all the samples in your dataset are the same length and no padding is necessary, you can skip this argument.
If you need to do something more complex than just padding samples (e.g. corrupting tokens for masked language
modelling), you can use the `collate_fn` argument instead to pass a function that will be called to transform the
list of samples into a batch and apply any preprocessing you want. See our
[examples](https://github.com/huggingface/transformers/tree/main/examples) or
[notebooks](https://huggingface.co/docs/transformers/notebooks) to see this approach in action.

</Tip>

> Once we've created a `tf.data.Dataset` we can simply compile and fit our model as before:

Reviewer 2: ```suggestion
Once you've created a `tf.data.Dataset`, you can compile and fit the model as before:
```
